dataLayer Builder + Enhanced Ecommerce.tpl

```diff
@@ -0,0 +1,554 @@
+          }
+        }
+      ]
+    },
+    "isRequired": true
+  },
+  {
+    "instance": {
+      "key": {
+        "publicId": "access_globals",
+        "versionId": "1"
+      },
+      "param": [
+        {
+          "key": "keys",
+          "value": {
+            "type": 2,
+            "listItem": [
+              {
+                "type": 3,
+                "mapKey": [
+                  {
+                    "type": 1,
+                    "string": "key"
+                  },
+                  {
+                    "type": 1,
+                    "string": "read"
+                  },
+                  {
+                    "type": 1,
+                    "string": "write"
+                  },
+                  {
+                    "type": 1,
+                    "string": "execute"
+                  }
+                ],
+                "mapValue": [
+                  {
+                    "type": 1,
+                    "string": "dataLayer"
+                  },
+                  {
+                    "type": 8,
+                    "boolean": true
+                  },
+                  {
+                    "type": 8,
+                    "boolean": true
+                  },
+                  {
+                    "type": 8,
+                    "boolean": false
+                  }
+                ]
+              }
+            ]
+          }
+        }
+      ]
+    },
+    "clientAnnotations": {
+      "isEditedByUser": true
+    },
+    "isRequired": true
+  }
+]
+
+
+___SANDBOXED_JS_FOR_WEB_TEMPLATE___
+
+const makeTableMap = require('makeTableMap');
+const query = require('queryPermission');
+const createQueue = require('createQueue');
+const dataLayerPush = createQueue('dataLayer');
+const log = require('logToConsole');
+const merge = function() {
+  const obj = {},
+        il = arguments.length;
+  let i = 0,
+      key;
+  for (; i < il; i++) {
+    for (key in arguments[i]) {
+      if (arguments[i].hasOwnProperty(key)) {
+        obj[key] = arguments[i][key];
+      }
+    }
+  }
+  return obj;
+};
+const event = {'event': data.event};
+const currencyParameter = data.currencyParameter;
+let actionFieldParameters = data.actionFieldParameters ? makeTableMap(data.actionFieldParameters, 'name', 'value') : {};
+const actionFieldDropDown = data.actionFieldDropDown;
+ if (actionFieldDropDown === true) {
+    actionFieldParameters = actionFieldParameters;
+ } else {
+   actionFieldParameters = undefined;
+ }
+
+const products = data.products;
+let customParameter = data.customParameter ? makeTableMap(data.customParameter, 'name', 'value') : {};
+const customParameterDropDown = data.customParameterDropDown;
+
+if (customParameterDropDown === true) {
+  customParameter = customParameter;
+} else {
+  customParameter = undefined;
+}
+//const customParameter = data.customParameter ? makeTableMap(data.customParameter, 'name', 'value') : {};
+const eventType = data.eventType;
+var dLayer = '';
```

Comment:
As a coding pattern this is awkward, as you're intitializing `dLayer` as something it doesn't end up being (an object).

So

```
const dLayer = {};
```

and then instead of reinitializing it in the upcoming if...else blocks, just set its ecommerce property:

```
dLayer.ecommerce = {...};
```

Reply:
Good observation, the devils in the details. Thanks for the pointers on how to clean up my coding and how to code more efficiently. :)